Issue template for adding a new experiment (#628)
Qiskit/qiskit-experiments: .github/ISSUE_TEMPLATE/NEW_EXPERIMENT.md

---
name: "\U0001F41B New experiment proposal"

> about: "Proposal and workflow guideline for adding a new experiment \U0001F914."

Author: The table and emojis are sort of a convention for the issue templates in various qiskit repos, I'm just following the tradition here. I did change the emojis to ones which are better for this issue (in my eyes)

title: ''
labels: enhancement
assignees: ''

---

<!-- ⚠️ If you do not respect this template, your issue will be closed -->
<!-- ⚠️ Make sure to browse the opened and closed issues -->

# New experiment proposal
Provide the experiment and implementation details below and ask the qiskit-experiments core team to review 
your proposal. 

## General details

### Experiment name
What is the experiment class name ? This name will also be used in the API documentation and tutorial.

### Experiment type
What type the experiment is ? Characterization, calibration, verification, validation, else? 

### Experiment protocol
Provide a concise description of the experiment. Make sure you cover the following aspects:  
* What is the main goal of the experiment ?

> * Describe the circuits/pulses and their main parameters needed for the experiment

Author: Yes, I would keep it in the implementation details. Added the specific ones you suggest here

* Input parameters, required and optional, and types of each
* Experiment options and default values
* Are there any limitations or special requirements for running the experiment run as part of a composite experiments? 
Provide details.

### Experiment analysis
* Analysis options and default values
* How does it generate initial guesses
* What plots will be generated ? 

---

# Workflow
These are the steps required for adding a new experiment to qiskit-experiments. It is advisable to follow this workflow
in order and get the proper review approvals from the qiskit-experiments core team before moving on to subsequent steps 
in the workflow. 

NOTE: An experiment PR failing to complete all the workflow steps below will not be merged !

- [ ] Complete the proposal section above and have your proposal reviewed and approved 
- [ ] Open a PR and link it to the new experiment issue
- [ ] Create the main module files (at least one for the experiment and one for its analysis). Place them in the right 
folders 
- [ ] Define and implement the experiments APIs 
- [ ] Have the experiment API reviewed and approved
- [ ] Implement all the experiment and analysis functionality

> - [ ] Add unit testing for the experiment and analysis classes. If needed implement a mock-backend for your experiment

Author: We should probably have a thorough testing specification written somewhere (e.g. under the test folder) which can be linked to from different guides, including this one. I think what we have now is a good start and we can add more details later on.

- [ ] Ask for a detailed implementation review
- [ ] Verify that your experiment runs properly on a real device and that the results make sense

> - [ ] Verify that your experiment runs properly in the context of a parallel experiment, where sub-experiments run on 

Author: Added a statement about covering `ParallelExperiment` context in testing

different qubits
- [ ] Verify aggregated cases, where `run` is executed with the `experiment_data` parameter set to an existing experiment 
data

> - [ ] Verify that figures look OK: for regular experiments, parallel experiments, aggregated experiments

Author: done

- [ ] Write API docs for all your API methods. Follow the guideline [here](https://github.com/Qiskit/qiskit-experiments/blob/main/CONTRIBUTING.md)
- [ ] Write a tutorial for your experiment. Follow the guideline [here](https://github.com/Qiskit/qiskit-experiments/blob/main/docs/tutorials/GUIDELINES.md) 

> - [ ] Ask for a final review of the documentation and implementation

Author: Right, added.

---
name: "\U0001F4CB New experiment proposal"
about: "Proposal and workflow guideline for adding a new experiment \U0001F52C."
title: ''
labels: enhancement
assignees: ''

---

<!-- ⚠️ If you do not respect this template, your issue will be closed -->
<!-- ⚠️ Make sure to browse the opened and closed issues -->

# New experiment proposal

> Provide the experiment and implementation details below and ask the qiskit-experiments core team to review 

Reviewer: Most of the text within these sections are reminders for the author. I wonder if they would make more sense inside of `<!-- ... -->` tags so that they are visible to the author while writing the issue but the author does not need to delete all of them before submitting the issue.

Author: good idea, implemented

* What are the main outputs of the experiment?

### Experiment analysis
Provide a concise description of the experiment's analysis. Make sure you cover the following aspects:
* What is the main fit model for the experiment? 
* What are the main fit parameters? Specify parameter defaults and bounds where relevant
* How would you evaluate whether the fit is good or bad?

### References 
If applicable, provide a few of the most relevant references here.

## Implementation details
Provide additional details pertaining to the implementation of the experiment and its analysis. Use the following as 
guiding points. Provide code snippets and usage examples where appropriate. 
### Experiment implementation
* What are the base classes for the experiment and its analysis?
* List the input parameters, required and optional, and types of each
* Experiment options and default values (e.g. default transpile and run options)
* Are there any limitations or special requirements for running the experiment as part of a composite experiment? 
Provide details.

### Experiment analysis
* Analysis options and default values
* How does it generate initial guesses
* What plots will be generated? 

---

> # Workflow

Reviewer: Basically everything in this section relates to the PR that the author needs to submit implementing the experiment after the proposal is approved. Perhaps this section could be moved to a "New Experiment Pull Request" template instead of this new experiment  issue template. If you follow this suggestion, I would link to the pull request template here at the bottom of the issue template.

Author: I think there is an advantage in having everything in one place. One of those being that the authors would see right away all the requirements for adding a new experiment. And they can obviously copy the task list to their PR. I'd say let's keep it this way and we can modify later if it turns out to be a hassle or not being followed in the PRs

These are the steps required for adding a new experiment to qiskit-experiments. It is advisable to follow this workflow
in order and get the proper review approvals from the qiskit-experiments core team before moving on to subsequent steps 
in the workflow. 

NOTE: An experiment PR failing to complete all the workflow steps below will not be merged!

> - [ ] Complete the proposal section above and have your proposal reviewed and approved 

Reviewer: Since the proposal is an issue, what is the approval process? How does the author know it is safe to begin implementing the experiment? Maybe adding a `experiment-approved` tag to the issue (currently, I think anyone can add tags to issues but I see that there is a beta feature for protected tags that only maintainers can add).

Author: I assumed that once the author completes the proposal description, he or she will request a review (as stated in the first step here). Then there will probably be some interchange with the package reviewers at the end of which they will say something like 'looks good, let's have this experiment added'. Adding an `approved` label would be nice addition, agreed. Do you think we should explicitly write that the authors should make sure they get such labeling?

Reviewer: It is one idea. There might be another way. I think it would be nice to have something that is clear since in an issue discussion there could be some ambiguity.

- [ ] Open a PR and link it to the new experiment issue
- [ ] Create the main module files (at least one for the experiment and one for its analysis). Place them in the right 
folders 
- [ ] Define and implement the experiments APIs 
- [ ] Have the experiment API reviewed and approved
- [ ] Implement all the experiment and analysis functionality

> - [ ] Verify that your experiment runs properly on a real device and that the results make sense

Reviewer: Everything from here down could double as a checklist for a reviewer (though maybe there are additional steps?).

Author: Agree. We will later add a checklist for reviewers which will include many of the steps here for checking

---
name: "\U0001F4CB New experiment proposal"
about: "Proposal and workflow guideline for adding a new experiment \U0001F52C."
title: ''
labels: enhancement
assignees: ''

---

<!-- ⚠️ If you do not respect this template, your issue will be closed -->
<!-- ⚠️ Make sure to browse the opened and closed issues -->

# New experiment proposal
Provide the experiment and implementation details below and ask the qiskit-experiments core team to review 
your proposal. 

## General details

### Experiment name
What is the experiment class name? This name will also be used in the API documentation and tutorial.

### Experiment type

> What type the experiment is? Characterization, calibration, verification, validation, else? 

Reviewer 2: ```suggestion
What is the experiment type? Characterization, calibration, verification, validation, or other? 
```

### Experiment protocol
Provide a concise description of the experiment. Make sure you cover the following aspects:  
* What is the main goal of the experiment?
* Describe the circuits/pulses and their main parameters needed for the experiment
* What are the main outputs of the experiment?

### Experiment analysis
Provide a concise description of the experiment's analysis. Make sure you cover the following aspects:
* What is the main fit model for the experiment? 

> * What are the main fit parameters?

Reviewer 2: Maybe also add defaults and bounds for the parameters, if relevant?

Author: added

* How would you evaluate whether the fit is good or bad?

### References 
If applicable, provide a few of the most relevant references here.

## Implementation details
Provide additional details pertaining to the implementation of the experiment and its analysis. Use the following as 
guiding points. Provide code snippets and usage examples where appropriate. 
### Experiment implementation
* What are the base classes for the experiment and its analysis?
* List the input parameters, required and optional, and types of each
* Experiment options and default values (e.g. default transpile and run options)
* Are there any limitations or special requirements for running the experiment as part of a composite experiment? 
Provide details.

### Experiment analysis
* Analysis options and default values
* How does it generate initial guesses
* What plots will be generated? 

---

# Workflow
These are the steps required for adding a new experiment to qiskit-experiments. It is advisable to follow this workflow
in order and get the proper review approvals from the qiskit-experiments core team before moving on to subsequent steps 
in the workflow. 

> NOTE: An experiment PR failing to complete all the workflow steps below will not be merged !

Reviewer 2: ```suggestion
NOTE: An experiment PR failing to complete all the workflow steps below will not be merged!
```

- [ ] Complete the proposal section above and have your proposal reviewed and approved 
- [ ] Open a PR and link it to the new experiment issue
- [ ] Create the main module files (at least one for the experiment and one for its analysis). Place them in the right 
folders 
- [ ] Define and implement the experiments APIs 

> - [ ] Have the experiment API reviewed and approved

Reviewer 2: Do we want people to get several stages of reviews for the PR? I'm wondering if we have enough reviewer bandwidth. Should the API review and approval be part of the proposal?

Author: The idea was to avoid the situation where developers will have to revert too much of their work because of certain development milestones (such as the API defs) being reviewed late. Being rather short, I thought API review can be done with relatively small effort. How about we merge the implementation and final review stages to one stage only? This leaves us with API review and final review (which includes implementation, testing and doc review)

Reviewer 2: That sounds good to me.

Author: done

Reviewer: Similar to my question above -- should we use an `api-approved` tag to PR's that have had the API approved, or what is the approval process? By the way, what are you thinking about regarding an API? Most experiments do not really define an API. They just implement or override methods of `BaseExperiment` and `BaseAnalysis`.

different qubits. Verify this also on a real device
- [ ] Verify that figures look OK: for regular experiments, parallel experiments, aggregated experiments
- [ ] Verify that experiment data is properly saved to and load from the results DB (experiments service), and that your
experiment data is displayed correctly in the results DB webpage. 
- [ ] Add unit testing for the experiment and analysis classes. If needed implement a mock-backend for your experiment. 
Include in your testing running the experiment in the context of `ParallelExperiment`
- [ ] Ask for a detailed implementation review
- [ ] Write API docs for all your API methods. Follow the guideline [here](https://github.com/Qiskit/qiskit-experiments/blob/main/CONTRIBUTING.md)
- [ ] Write a tutorial for your experiment. Follow the guideline [here](https://github.com/Qiskit/qiskit-experiments/blob/main/docs/tutorials/GUIDELINES.md)
- [ ] Add a new release note. Follow the guideline [here](https://github.com/Qiskit/qiskit-experiments/blob/main/CONTRIBUTING.md#adding-a-new-release-note) 
- [ ] Ask for a final review of the documentation and implementation

> - [ ] Celebrate !

Reviewer 2: ```suggestion
- [ ] Celebrate!
```
